### DISCUSSION TITLE
PyQt6 not importing correctly

### DISCUSSION THREAD

**Discussion by need4swede** [pyinstaller/pyinstaller#6851] [Help] [answered]
```
MacOS Monterey (12.3.1)
M1 Air (arm64)

pyinstaller==5.1
PyQt6==6.3.0
Python==3.9.12
```

I have a PyQt6 project that I am trying to package as an app. I used "--onefile --windowed" AND "--onedir --windowed", respectively, to accomplish this.

It crashes immediately upon launch when running the *.app ; So I tried running it via the *.exe and got the following:

> ImportError: dlopen(/Users/need4swede/myapp/dist/app/PyQt6/QtWidgets.abi3.so, 0x0002): Symbol not found: __ZNK11QHeaderView15initStyleOptionEP17QStyleOptionFrame

I thought that was odd. So I ran pyinstaller without arguments and went inside and replaced the 'PyQt6' folder with the one from my python packages. Voila - I can launch the app through the executable. So I thought "no problem, I'll just do --add-data {...PyQt6:PyQt6} whatever and it will be fixed, right?"

So I did that, and now I'm getting this:

> qt.qpa.plugin: Could not load the Qt platform plugin "cocoa" in "" even though it was found.
> This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
> 
> Available platform plugins are: cocoa, minimal, offscreen.

I don't get it. Replacing the PyQt6 folder fixes it if I do it after pyinstaller, but if tell pyinstaller to use my directory, I get this message? I'm not an expert at this, so any help would be appreciated. Ideally would like to have my app launch without all the hassle.

**Comment by rokm** [comment, marked as answer]
Do you have `PyQt5` installed in your environment as well? Can you check your `build/<name>/Analysis-00.toc` file to see where `QtWidgets` shared library (the one whose entry looks like `('QtWidgets', 'source_path', 'BINARY')`) was collected from? Was it from `PyQt5` package instead of `PyQt6` one?

Are you importing `PyQt6` directly or are you using a higher level abstraction library (like `QtPy`)? If you are using `QtPy` or equivalent, you need to either ensure that only one Qt package is installed in your build environment, or suppress the unused ones via `pyinstaller --exclude PyQt5 --exclude PySide2 --exclude PySide6 [...]`. Otherwise, PyInstaller will collect all of them, and chaos will ensue...

**Reply by need4swede** [reply]
I do have PyQt5!

I tried doing this in a virtual env with no PyQt5 installed, but still got the same error. 
But alas, I do see PyQt5 in the build regardless, so that may be the issue!

```
('PyQt6.QtWidgets',
   '/opt/homebrew/lib/python3.9/site-packages/PyQt6/QtWidgets.abi3.so',
   'EXTENSION')

('PyQt5.QtWidgets',
   '/opt/homebrew/lib/python3.9/site-packages/PyQt5/QtWidgets.abi3.so',
   'EXTENSION')
```
Both of these buggers in the .toc!
PyQt5 comes after PyQt6.

> Are you importing PyQt6 directly or are you using a higher level abstraction library (like QtPy)?

Are you asking how I import PyQt6 into my python project?
Like this,
```
from PyQt6.QtWidgets import QApplication...
from PyQt6.QtGui import QIcon...
from PyQt6.QtCore import Qt, QDir, QCoreApplication...
```

I'll try passing '--exclude PyQt5' and will update with what happens!
Thanks a bunch for your help!

-----------------

**EDIT: You're amazing! It works! --exclude PyQt5 fixed it :)**

--onefile --windowed doesn't seem to work, though. How can I go about debugging that?
--onedir --windowed works just fine, though!

Thank you so much!

**Reply by rokm** [reply]
> --onefile --windowed doesn't seem to work, though. How can I go about debugging that?

Are you using `QtWebEngine` by any chance? That won't work in `onefile` on macOS (and currently cannot be fixed).

Otherwise, without seeing the code, the only advice I can give you is to wrap your whole entry-point script (i.e., the .py file you are passing to PyInstaller; and this means *all* code, including the initial imports) in a try/except block and write the exception message to a file at fixed location (e.g., hardcoded path to your user's home directory). That will give you a starting point to investigate what's going on.

Alternatively, does it work with `--onefile` but without `--windowed` (or with `--windowed`, but by running the `dist/<program>` instead of running the app bundle `dist/<program>.app`?

**Reply by need4swede** [reply]
Here are all the imports of the app:

```
import sys, subprocess, warnings
import os, platform, requests, re, time, base64, shutil
import pandas as pd
from pathlib import Path
from PyQt6.QtWidgets import QApplication, QStatusBar, QWidget, QLabel, QLineEdit, QPushButton, QProgressBar, QMessageBox, QFileDialog, QVBoxLayout, QHBoxLayout, QCheckBox
from PyQt6.QtGui import QIcon, QCursor, QFont
from PyQt6.QtCore import Qt, QDir, QCoreApplication
from bs4 import BeautifulSoup
```

That's a good idea for the debug! I'll give that a go and see what happens!

I tried `--onefile` without `--windowed`, and it works!
`--onedir --windowed` works
`--onefile `works
`--onefile --windowed` does not work

I'll try that try/except trick and see what's what!

### EDIT: 
That's odd, `--onefile --windowed` works now, albeit strangely. The app starts, then it appears to close right away...and a few seconds later launches again. And it appears to be running. This doesn't happen with the executable, or when doing `--onedir --windowed` My error.txt file has nothing written in it, though, as if the exception was never thrown. 

I also don't understand the distinction between these two. I deleted the directory that accompanied my `--onedir --windowed` install, and the *.app still runs just fine. In fact, this is working exactly how I wanted it to work. Is it suppose to be this way?

**Reply by rokm** [reply]
> That's odd, `--onefile --windowed` works now, albeit strangely. The app starts, then it appears to close right away...and a few seconds later launches again. And it appears to be running. This doesn't happen with the executable, or when doing `--onedir --windowed` My error.txt file has nothing written in it, though, as if the exception was never thrown.
> 
> I also don't understand the distinction between these two. I deleted the directory that accompanied my `--onedir --windowed` install, and the *.app still runs just fine. In fact, this is working exactly how I wanted it to work. Is it suppose to be this way?

Yeah, that's normal for `onefile` build. Because it consists of two processes - the parent unpacks the contents of embedded archive, and then launches the child, which executes your frozen python code (in contrast, onedir runs your frozen python code directly in its main/only process). When ran as the app bundle, each process gets its own icon; but we put the parent process to the background in an attempt to hide it. This doesn't happen immediately, so the end result is that it looks like the application crashed (the parent process is finally put to background and its icon disappears) and then re-launches (the child process is finally started and gets its own icon).

In onedir mode, the shared libraries and data files are collected into application's directory, whereas in onefile, they are collected into archive that's embedded into executable (and therefore needs to be unpacked on each run). Using --windowed flag generates both a "regular" POSIX build (console-enabled executable), e.g. `dist/<program>`, and an .app bundle, e.g. `dist/<program>.app`. So together with --onedir/--onefile you can actually have four combinations: onedir POSIX build, onefile POSIX build, onedir POSIX build + onedir app bundle, onefile POSIX build + onefile app bundle).

On macOS, if you want to create an app bundle and plan to notarize it, you will likely need to go with a onedir build.

**Reply by need4swede** [reply]
Wow, that's super informative. I'm very new to all this and understood that perfectly - very well put. Your feedback and help has been nothing short of incredible. Thank you very much!
